### ISSUE TITLE
Proposal: "object-curly-newline" add configuration for named imports and exports (or create new rule for import/export specifiers)

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#8501]
I'm creating a new issue for this because I'd like to see it considered for inclusion.

From the original post (see #7319):

> ESLint version `v3.5.0`
> 
> Rule [object-curly-newline](http://eslint.org/docs/rules/object-curly-newline) currently supports:
> - `"ObjectExpression"` configuration for object literals;
> - `"ObjectPattern"` configuration for object patterns of destructuring assignments.
> 
> ## Proposal
> 
> I propose to add similar configuration for named imports and exports:
> 
> ``` js
> {
>     "object-curly-newline": ["error", {
>         "ImportDeclaration": { "multiline": true },
>         "ExportDeclaration": { "multiline": true }
>     }]
> }
> ```
> 
> Possible values:
> ### always
> 
> Examples of **incorrect** code for this rule with the `"always"` option:
> 
> ``` js
> import {foo, bar} from 'foo-bar';
> import {foo as f, bar} from 'foo-bar';
> import {foo,
>     bar} from 'foo-bar';
> 
> export {foo, bar};
> export {foo as f, bar};
> export {foo,
>    bar};
> ```
> 
> Examples of **correct** code for this rule with the `"always"` option:
> 
> ``` js
> import {
>     foo,
>     bar
> } from 'foo-bar';
> import {
>     foo as f,
>     bar
> } from 'foo-bar';
> 
> export {
>   foo,
>   bar
> };
> export {
>   foo as f,
>   bar
> };
> ```
> 
> ### never
> 
> Examples of **incorrect** code for this rule with the `"never"` option:
> 
> ``` js
> import {
>     foo,
>     bar
> } from 'foo-bar';
> 
> export {
>    foo,
>    bar
> };
> ```
> 
> Examples of **correct** code for this rule with the `"never"` option:
> 
> ``` js
> import {foo, bar} from 'foo-bar';
> 
> export {foo, bar};
> ```
> 
> ### multiline
> 
> Examples of **incorrect** code for this rule with the default `{ "multiline": true }` option:
> 
> ``` js
> import {
>     foo, bar} from 'foo-bar';
> import {foo, bar
>     } from 'foo-bar';
> import {
>     foo, bar
> } from 'foo-bar';
> 
> export {
>    foo, bar};
> export {
>    foo, bar
> };
> ```
> 
> Examples of **correct** code for this rule with the default `{ "multiline": true }` option:
> 
> ``` js
> import {foo, bar} from 'foo-bar';
> import {foo as f, bar} from 'foo-bar';
> import {
>     foo,
>     bar
> } from 'foo-bar';
> import {
>     foo as f,
>     bar
> } from 'foo-bar';
> 
> export {
>    foo,
>    bar
> };
> ```
> 
> ### minProperties
> 
> Similar to current [`minProperties` behavior](http://eslint.org/docs/rules/object-curly-newline#minproperties). Maybe this option should be renamed to `minImportSpecifiers` or similar.
> 
> ## Question
> 
> Maybe this should be separate rule instead of configuration of `object-curly-newline` rule.

Regarding the above question, I like the idea of augmenting object-curly-newline if other object-related rules tend to apply to import/export specifiers, but it's also true that some elements of objects (e.g., keys/values) simply don't apply to import specifiers. So I'm okay with a new rule as well. Hopefully that will be a quick bikeshed.

The last issue had been :+1:'d by the following people:
* @kaicataldo
* @not-an-aardvark
* @platinumazure

But we never had a champion. I'd like to see if we can get one this time. (I'm considering it. Just need to figure out my schedule.)

**Comment by evilebottnawi** [comment]
After accept this i can do PR :smile:

**Comment by platinumazure** [comment]
@kaicataldo @not-an-aardvark Is this something you would still be willing to :+1:?

**Comment by platinumazure** [comment]
I'll champion this, now this is accepted.

@eslint/eslint-team @evilebottnawi Any thoughts on whether this should be implemented as a new rule or as an extension of `object-curly-newline`?

**Comment by evilebottnawi** [comment]
@platinumazure for me in `object-curly-newline` :smile:

**Comment by kaicataldo** [comment]
From the last issue:
> I'm for the ability to enforce these styles, but I do think it's a little strange to have named `import` and `export` statements covered by this rule since they're not really objects. That being said, we do have precedence for this, as we currently check named `import` and `export` statements in [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing).

**Comment by evilebottnawi** [comment]
@kaicataldo maybe it be good do in `v4`? Anyway i do PR, rename is not long procedure :smile:

**Comment by kaicataldo** [comment]
@evilebottnawi Awesome! Yeah, I don't think renaming should be a blocker. Let's just add it to `object-curly-spacing` now and deal with renaming afterwards.

**Comment by kaicataldo** [comment]
I think we should decouple this discussion from idea of renaming this rule, though I still support doing that for clarity (in a backwards compatible way).

**Comment by OlegLustenko** [comment]
This something which my team will be happy to get

**Comment by jouncealimb** [comment]
I'd like to work on this!

**Comment by megalithic** [comment]
just discovering this issue and i concur this would be a wonderful rule change. just +1'ing from the community of users.

**Comment by jouncealimb** [comment]
Hi, sorry. I have not had as much time to work on this as i thought. i would be able to get to this really in 2 weeks from now, but for fairness, I will let anyone else take this on during this time. If no one else wants to after 2 weeks, I will pick it up again. Sorry for the hold up.

**Comment by nicholaschuayunzhi** [comment]
Hi @jouncealimb are you still working on this? Would like to give this a try.

**Comment by jouncealimb** [comment]
Hi @nicholaschuayunzhi, please feel free
